### PR TITLE
Fix #1347 - Back-end for sending email to invite users to host team using URL.

### LINKED ISSUE #1347: Add feature to invite users to host team using URL
Very similar to #1346. 

## Deliverables

Add a feature to invite new team member to a participant team using a Unique URL. Also, don't forget to write unit tests. :) 

## Steps

1. **user1** enters the email of his teammate (let's say **user2**)
2. Send an email to **user2** (if he/she is registered on EvalAI) with a unique URL
3. **user2** clicks on the URL to accept the invitation to the team.
4. *user1* gets an email saying that "**user2** has accepted your invite to team <blah>" (we can iterate over the email message later) 

### Some important points to note

- [ ] Checks performed before inviting to the team should be similar to current invitation logic

### PR DESCRIPTION
**Pull request by isht3**
* Define URL End-Points.

    * Define unique string encoder and decoder.\

    * Edit failing tests and add new tests for features.

Inviting.

Receiving E-mail.

Invited User Accepting.

Fixes #1347

### CODE REVIEW

**Comment by isht3** [comment]
@deshraj @RishabhJain2018 Could you tell me which code is not being covered? I believe I wrote tests to cover the new features.

File: apps/hosts/views.py [hunk 1]
```diff
@@ -13,10 +18,32 @@
                      ChallengeHostTeam,)
 from .serializers import (ChallengeHostSerializer,
                           ChallengeHostTeamSerializer,
-                          InviteHostToTeamSerializer,
                           HostTeamDetailSerializer,)
 
 
+SECRET_KEY = "hotchocolates"
```
Comment (deshraj) [inline apps/hosts/views.py]: We would ideally want to use the SECRET_KEY in the django project settings.

File: apps/hosts/views.py [hunk 2]
```diff
@@ -13,10 +18,32 @@
...
+
+
+def ENCODE_DATA(data):
```
Comment (deshraj) [inline apps/hosts/views.py]: Please change the method name to `encode_data()` and you should move this method to `utils.py` in the same directory.

File: apps/hosts/views.py [hunk 3]
```diff
@@ -13,10 +18,32 @@
...
+    """
+    Turn `data` into a hash and an encoded string, suitable for use with `DECODE_DATA`.
+    """
+    encoded = []
+    for i in data:
+        encoded.append(base64.encodestring(i).split("=")[0])
+    return encoded
+
+
+def DECODE_DATA(data):
```
Comment (deshraj) [inline apps/hosts/views.py]: Similar comment as for `encode_data()`

File: apps/hosts/views.py [hunk 4]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
 @api_view(['POST'])
 @permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
 @authentication_classes((ExpiringTokenAuthentication,))
-def invite_host_to_team(request, pk):
-
+def email_invite_host_to_team(request, pk):
```
Comment (deshraj) [inline apps/hosts/views.py]: Please add docstring.

File: apps/hosts/views.py [hunk 5]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+    except User.DoesNotExist:
+        response_data = {'error': 'User does not exist with this email address!'}
+        return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+    if email == request.user.email:
+        response_data = {'error': 'A host cannot invite himself'}
+        return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+    team_hash, email_hash = ENCODE_DATA([str(pk), email])
+    unique_hash = (team_hash + "/" + email_hash)
```
Comment (deshraj) [inline apps/hosts/views.py]: Use String formatting using `.format()` instead of using `+`

File: apps/hosts/views.py [hunk 6]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+    url = request.data['url'].split("team")[0]
+    full_url = url + "participant-invitation/" + unique_hash
+    team_name = challenge_host_team.team_name
+    body = "You've been invited to join the host team {}. " \
```
Comment (deshraj) [inline apps/hosts/views.py]: Store the email body in `hosts/templates` similar to how django allauth is storing. I think I have shared the link with you earlier. Let me know if I haven't done that.

File: apps/hosts/views.py [hunk 7]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+           "Click the bottom link to accept the " \
+           "invitation and to participate in challenges. \n{}"
+    message = body.format(team_name, full_url)
+    subject = "You have been invited to join the {} host team at CloudCV!".format(team_name)
+    try:
+        send_mail(subject,
+                  message,
+                  "admin@CloudCV.com",
```
Comment (deshraj) [inline apps/hosts/views.py]: Create a field to store the email id in settings and fetch from there.

File: apps/hosts/views.py [hunk 8]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+                  [email],
+                  fail_silently=False,)
+        response_message = "{} has been invited to join the host team {}".format(email, team_name)
```
Comment (deshraj) [inline apps/hosts/views.py]: change `response_data` to `response`

File: apps/hosts/views.py [hunk 9]
```diff
@@ -212,20 +239,63 @@ def remove_self_from_challenge_host_team(request, challenge_host_team_pk):
...
+    except BadHeaderError:
+        response_data = {'error': 'There was some error while sending the invite.'}
+        return Response(response_data, status=status.HTTP_417_EXPECTATION_FAILED)
+
+
+@throttle_classes([UserRateThrottle])
+@api_view(['POST'])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((ExpiringTokenAuthentication,))
+def invitation_accepted(request, team_hash, email_hash):
```
Comment (deshraj) [inline apps/hosts/views.py]: Add docstring

File: apps/hosts/views.py (hunk 4 above)
Comment (deshraj) [reply apps/hosts/views.py]: I think you are missing bunch of conditions in this method. Let's say the new member is already a participant for a challenge which is hosted by the challenge host team where he is invited then he should not be able to join the team.

File: tests/unit/hosts/test_urls.py [hunk 10]
```diff
@@ -51,6 +56,12 @@ def test_host_urls(self):
...
 
-        url = reverse_lazy('hosts:invite_host_to_team',
+        url = reverse_lazy('hosts:email_invite_host_to_team',
                            kwargs={'pk': self.challenge_host_team.pk})
         self.assertEqual(url, '/api/hosts/challenge_host_teams/' + str(self.challenge_host_team.pk) + '/invite')
+
+        url = reverse_lazy('hosts:host_invitation_accepted',
+                           kwargs={'team_hash': self.team_hash.strip("\n"),
+                                   'email_hash': self.email_hash.strip("\n")})
+        self.assertEqual(url, '/api/hosts/invitation_host_team/' + self.team_hash.strip("\n") + '/'
```
Comment (deshraj) [inline tests/unit/hosts/test_urls.py]: Please use `.format()`

**Comment by deshraj** [comment]
Overall looks good. Just need minor tweaks and then it will go through easily.

File: apps/hosts/serializers.py [hunk 11]
```diff
@@ -42,34 +42,6 @@ class Meta:
         fields = ('id', 'user', 'team_name', 'status', 'permissions',)
 
 
-class InviteHostToTeamSerializer(serializers.Serializer):
```
Comment (deshraj) [inline apps/hosts/serializers.py]: I am not sure why this part of code is removed. Am I missing something here?

Comment (isht3) [reply apps/hosts/serializers.py]: Initially, the E-mail and Id would be in the request, but here, we're decoding the data from the URL. So the serializer needed a major overhaul and I found this method to be much efficient than re-writing the serializer.

File: apps/hosts/views.py (hunk 6 above)
Comment (isht3) [reply apps/hosts/views.py]: No, I think you didn't. How do you access the template from the view?

File: apps/hosts/views.py (hunk 4 above)
Comment (isht3) [reply apps/hosts/views.py]: @deshraj I've added it.

**Comment by isht3** [comment]
@deshraj Only transferring the E-mail to the template is left. Could you share the link you mentioned?

**Comment by deshraj** [comment]
@isht3: while trying to review, I am getting this error:

```
KeyError at /api/hosts/challenge_host_teams/1/invite
'url'

Request Method: POST
Request URL: http://localhost:8000/api/hosts/challenge_host_teams/1/invite
Django Version: 1.10.2
.
.
.
```
 Please let me know once you have fixed this.

**Comment by deshraj** [comment]
@isht3: It looks like we are converging on this. I can see the email content but when I try to click on the link to accept the invite then it shows a 404 page. Please see the screenshot below:

screen shot 2018-02-28 at 7 49 41 pm

**Comment by isht3** [comment]
@deshraj That is because this is only the back-end implementation of the feature. The front-end part for this feature is in the PR #1488. 😄

**Comment by isht3** [comment]
@deshraj Similar implementation in PR #1414 #1415

**Comment by deshraj** [comment]
> @deshraj That is because this is only the back-end implementation of the feature. The front-end part for this feature is in the PR #1488. 😄

I actually merged both the pull requests in one branch and then tried to do it. Can you please check it on your end again?

**Comment by isht3** [comment]
@deshraj This is the first time I'm getting this error, any idea what may be wrong?
```
 FAILED TESTS:
  Unit Tests for auth controller
    _ "before each" hook: WorkFn for "has default values"
      PhantomJS 2.1.1 (Linux 0.0.0)
    undefined is not a function (evaluating '$window.ga('create', 'UA-45466017-2', 'auto')')
    runFunc@frontend/src/js/route-config/route-config.js:498:19
    invoke@frontend/dist/vendors/vendor.js:49:395
    frontend/dist/vendors/vendor.js:51:316
    r@frontend/dist/vendors/vendor.js:13:11
    gb@frontend/dist/vendors/vendor.js:51:290
    WorkFn@node_modules/angular-mocks/angular-mocks.js:3173:60
```

**Comment by spyshiv** [comment]
@isht3 @deshraj There is 404 error since url is not valid. Acc. to route config of PR #1488 the expected url should be `http://127.0.0.1:8888/web/challenge-host-invitation/:team_hash/:email_hash` instead of `http://127.0.0.1:8888/web/challenge-host-participant-invitation/:team_hash/:email_hash`
`

**Comment by spyshiv** [comment]
@isht3 How to reproduce unit test error you mentioned in your previous message?

**Comment by isht3** [comment]
@spyshiv, I merged the contents of the back-end and front-end together and tried running it. I'll check again today. Any ideas on what may be wrong?

**Comment by deshraj** [comment]
@isht3 when you merged, then is it working properly on your system? 

Also, I think that when you copied the logic then you forgot to change the URL for the host invitation. See @spyshiv's message: https://github.com/Cloud-CV/EvalAI/pull/1489#issuecomment-369690932

**Comment by isht3** [comment]
@deshraj I think that was the issue, I resolved it. Can you check again? :smile:

**Comment by isht3** [comment]
@deshraj Can you please review this and #1414 .

**Comment by RishabhJain2018** [comment]
Closing this due to inactivity.

**Pull request closed without merging**